### PR TITLE
Rework the WASM code

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The objective of this PR is to:

- Make WASM build produce an expected package.
- Investigate if we can share BitwardenClient between WASM and NAPI.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -21,3 +21,8 @@ codegen-units = 1
 # Stripping the binary reduces the size by ~30%, but the stacktraces won't be usable anymore. 
 # This is fine as long as we don't have any unhandled panics, but let's keep it disabled for now
 # strip = true
+
```
Comment (dani-garcia) [inline Cargo.toml:24]: This profile doesn't seem used anywhere? We can replace `--release` for `--profile release-wasm` in the README instructions if it's useful

File: README.md [hunk 2]
```diff
@@ -91,3 +91,35 @@ generation. In that case we excluded the file from being generated and manually
...
+Install `wasm32-unknown-unknown` rust target.
+
+```bash
+rustup target add wasm32-unknown-unknown
+```
+
+Install wasm-bindgen-cli
+
+```bash
+cargo install -f wasm-bindgen-cli
```
Comment (dani-garcia) [inline README.md:106]: I had to separately install `wasm-opt`, used `brew install binaryen` on Mac, or from the releases page: https://github.com/WebAssembly/binaryen/releases. Probably a good idea to mention it here as well

File: crates/bitwarden/Cargo.toml [hunk 3]
```diff
@@ -19,40 +19,46 @@ secrets = []  # Secrets manager API
 internal = [] # Internal testing methods
 mobile = []   # Mobile-specific features
 
+stdweb = ["instant/stdweb"]
```
Comment (dani-garcia) [inline crates/bitwarden/Cargo.toml:22]: This seems unused as well.

**Review by dani-garcia — COMMENTED** [review]
Some small comments, we're also now using `chrono` to handle time serialization, I think it would be a good idea to also use it where we are using `std::time::Instant`. It should be WASM compatible as well and we won't need a separate `instant` library just for this use case. I can submit a PR with just that change you want.

**Review by vgrassia — APPROVED** [review]

File: crates/bitwarden/Cargo.toml [hunk 4]
```diff
@@ -37,6 +39,7 @@ chrono = { version = ">=0.4.26, <0.5", features = [
 getrandom = { version = ">=0.2.9", features = ["js"] }
 hkdf = ">=0.12.3, <0.13"
 hmac = ">=0.12.1, <0.13"
+instant = ">=0.1.12, <0.2"
```
Comment (dani-garcia) [inline crates/bitwarden/Cargo.toml:42]: Now that we changed the token renewal instants to timestamps this dependency isn't needed anymore, but we probably need to add the `wasmbind` feature to `chrono`

File: README.md [hunk 5]
```diff
@@ -97,3 +97,39 @@ generation. In that case we excluded the file from being generated and manually
...
+```bash
+# dev
+cargo build -p bitwarden-wasm -p bitwarden --target wasm32-unknown-unknown --features wasm-bindgen
+wasm-bindgen --target bundler --out-dir languages/js/wasm ./target/wasm32-unknown-unknown/debug/bitwarden_wasm.wasm
+wasm-bindgen --target nodejs --out-dir languages/js/wasm/node ./target/wasm32-unknown-unknown/debug/bitwarden_wasm.wasm
+
+# release
+cargo build -p bitwarden -p bitwarden-wasm --target wasm32-unknown-unknown --features wasm-bindgen --release
+wasm-bindgen --target bundler --out-dir languages/js/wasm ./target/wasm32-unknown-unknown/release/bitwarden_wasm.wasm
+wasm-bindgen --target nodejs --out-dir languages/js/wasm/node ./target/wasm32-unknown-unknown/release/bitwarden_wasm.wasm
```
Comment (dani-garcia) [inline README.md:126]: Are we going to support using the wasm builds in node while also offering the node-api support?

Comment (Hinton) [reply README.md:126]: I would like to offer it at least while we explore if we can transition off the napi bindings internally.

**Review by dani-garcia — DISMISSED** [review]

**Review by dani-garcia — APPROVED** [review]

**Review by vgrassia — APPROVED** [review]

**Pull request merged**